Consolidate transaction signature checking

ximinez opened the pull request:
The idea here is to clean up the wide variety of methodologies used to check transaction signatures, especially on receipt from:
*    Tx received from peer
*    Tx received from client
*    Tx signed and submitted locally.

Now, all checks aggressively go through a single function (`HashRouter::checkValidity`), and the `sigVerify` lambda callback has been eliminated from the `Transactor`.

Reviewers: @vinniefalco @miguelportilla. Should probably also be checked out by @JoelKatz when he is available.
*The first commit ("Devirtualize HashRouter") can be ignored*, since it's already covered by #1223.

Since I'm about to go on vacation, if there are any problems, either someone else can take ownership, or the PR can be closed until I get back.

miguelportilla commented:
Its beyond the scope of this PR but I think the hash router class can be improved. Calling any of its methods creates an Entry in the hash table. I would not expect methods such as getFlags or getDetails to do that and return default flags for the newly created Entry. In fact there is no way to determine if an Entry exists. I noticed at one point you started implementing methods that would allow you to retrieve an Entry or its flags without creating one but I suppose you ran into some issues with that.

vinniefalco commented:
@miguelportilla The names of the member functions are misleading, I agree. But before promoting a fix, I suggest auditing all callers to understand how the class is used.

miguelportilla commented:
This passes all tests and builds fine on all platforms. I left some comments but not holding back. 
:+1:

JoelKatz commented:
@miguelportilla The existence of an entry in the hash router means that you've seen the object corresponding to that entry recently. Since seeing an object is the only way to get the hash, if our code has the hash of an entry and is processing it, that entry should exist in the hash router. Under what circumstance would you be processing an entry but not want the hash router to have the state that reflects that you're processing that entry?

miguelportilla commented:
@JoelKatz Thank you, understood. I can't speak to circumstances as I don't know the calling code or assumptions. I had found it odd that a method to retrieve flags on a an object no present in the table would in turn create it and return default data, which could be misleading to programmer not aware of the assumptions.

ximinez commented:
@miguelportilla:
> Its beyond the scope of this PR but I think the hash router class can be improved. Calling any of its methods creates an Entry in the hash table. I would not expect methods such as getFlags or getDetails to do that and return default flags for the newly created Entry. In fact there is no way to determine if an Entry exists. I noticed at one point you started implementing methods that would allow you to retrieve an Entry or its flags without creating one but I suppose you ran into some issues with that.

Yeah, I implemented `const` versions of some of these functions in https://github.com/ximinez/rippled/commit/5954cf5b1bd0a99f0c002abb235cb129475e8985, because there were places that only examined the flags, and I wanted to replicated that behavior. I ended up removing them, because they were only used to set the `tapNO_CHECK_SIGN` flag, which evaporated when I moved signature checking out of the `Transactor`. When that happened, there was no need for the `const` functions anymore, either.

@JoelKatz provided a good explanation about why you always want to create the objects, but I did want to address
> Under what circumstance would you be processing an entry but not want the hash router to have the state that reflects that you're processing that entry?

I believe the answer to that question is if you do a read-only operation (`getFlags`), and don't do anything read-write until several seconds later (`setFlags`, `checkValidity`), the entry created by the read-only operation would expire from the cache significantly earlier than it would if we had waited for the read-write operation to create it. *However*, I do not believe that any such operations exist anymore, if they ever did.  If we're processing the object, we're not limiting ourselves to read-only operations. Because the cache has 1 second granularity, we can safely ignore the nano/micro/millisecond delay that exists now.

ximinez commented:
Pushed a couple more commits to address issues that were brought up in the comments.

JoelKatz commented on src/ripple/app/ledger/impl/LedgerConsensusImp.cpp:
```diff
@@ -1797,10 +1798,29 @@ applyTransaction (OpenView& view,
+        // signed. Any others will be rejected as malformed.
+        if (txn->isFieldPresent(sfAccount) &&
+            getApp().getHashRouter().checkValidity(*txn,
+#if RIPPLE_ENABLE_MULTI_SIGN
```
Since `Account` is `SOE_REQUIRED` in every transaction, the `isFieldPresent` test doesn't do anything. We need to make sure this code doesn't reject pseudo-transactions.

ximinez replied:
Great catch! I must have been in ~~deep~~ derp vacation brain. Fixed.

ximinez commented:
Since everybody has at least take a look at the baseline changes, I'm going to rebase against b5 now so I can get any conflicts resolved. I will hold off on squashing, though.

vinniefalco commented on src/ripple/app/ledger/impl/LedgerConsensusImp.cpp:
```diff
@@ -1175,10 +1176,15 @@ void LedgerConsensusImp::accept (std::shared_ptr<SHAMap> set)
-                getApp().getHashRouter().sigVerify(),
+        {
+            if (getApp().getHashRouter().checkValidity(
+                *item.second).first != HashRouter::SigBad)
```
We're not doing sig checking here while holding a lock, right?

ximinez replied:
We could be if the transaction isn't in the `HashRouter` or has expired, but we always were.

Before this change, we had to, because we had to keep the new open ledger away from other threads, and signature checking was in the Transactor. Now that it's not, we _could_ do the check before the lock is taken. However, I don't think it's worth it because the `localTx`s have already been checked, so the result is _likely_ to still be in `HashRouter`.

vinniefalco commented on src/ripple/app/ledger/impl/LedgerConsensusImp.cpp:
```diff
@@ -1826,10 +1828,30 @@ applyTransaction (OpenView& view,
+            true
+#else
+            flags & tapENABLE_TESTING
+#endif
```
I'm not really thrilled that the details of signature verification are scattered between various sources. Why does `LedgerConsensusImp.cpp` need to care about multi-signing? This is brittle.

ximinez replied:
As we discussed, these calls care because of testing. However, I did a quick audit of this particular call, and none of the callers can possibly have this flag set, so I'll change this to an assertion and use the default value.

vinniefalco commented on src/ripple/app/ledger/impl/LedgerConsensusImp.cpp:
```diff
@@ -1826,10 +1828,30 @@ applyTransaction (OpenView& view,
-                getConfig(), deprecatedLogs().
-                    journal("LedgerConsensus"));
+            getConfig(), deprecatedLogs().
+                journal("LedgerConsensus"));
```
Don't we have a `journal` object attached to `this`?

ximinez replied:
Nope. Doesn't look like it.

vinniefalco commented on src/ripple/app/ledger/impl/OpenLedger.cpp:
```diff
@@ -110,9 +110,15 @@ OpenLedger::accept(Rules const& rules,
+    {
+        if (router.checkValidity(
+            *item.second).first == HashRouter::SigBad)
+        {
```
I would leave the braces out but that's a style preference

ximinez replied:
I get a little nervous about readability when stuff is nested, but sure.

vinniefalco commented on src/ripple/app/ledger/impl/OpenLedger.cpp:
```diff
@@ -140,13 +146,29 @@ OpenLedger::apply_one (OpenView& view,
+    // signed. Any others will be rejected as malformed.
+    if (tx->isFieldPresent(sfAccount) &&
+        router.checkValidity(*tx,
+#if RIPPLE_ENABLE_MULTI_SIGN
```
Ugh, the details of multi-signing have infected my nice clean `OpenLedger` class?

ximinez replied:
None of the callers can possibly have this flag set either, so I'll change this to an assertion and use the default value here, too.

vinniefalco commented on src/ripple/app/misc/HashRouter.cpp:
```diff
@@ -26,124 +26,85 @@
 
     // See if any supressions need to be expired
-    std::map< int, std::list<uint256> >::iterator it = mSuppressionTimes.begin ();
+    auto it = mSuppressionTimes.begin ();
```
Beyond the scope of the PR but this looks like a candidate for `beast::aged_unordered_map`

vinniefalco commented on src/ripple/app/misc/HashRouter.cpp:
```diff
@@ -26,124 +26,85 @@
+    return emplace(index).first.getFlags ();
 }
 
 bool HashRouter::setFlags (uint256 const& index, int flags)
```
We're using `key` now instead of `index`. But that can be changed later.

ximinez replied:
Agreed. That would be a good task for a smaller, simpler PR.

vinniefalco commented on src/ripple/app/misc/HashRouter.cpp:
```diff
@@ -168,4 +128,66 @@ bool HashRouter::swapSet (uint256 const& index, std::set<PeerShortID>& peers, in
 }
 
+auto
+HashRouter::checkValidity(STTx const& tx, bool allowMultiSign)
```
Why does multisigning keep popping up?

vinniefalco replied:
So the `HashRouter` now knows about transactions? I don't like it

ximinez replied:
> Why does multisigning keep popping up?

Unit tests

> So the `HashRouter` now knows about transactions? I don't like it

Not anymore! `checkValidity` has been moved to apply.h/cpp.

vinniefalco commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -669,38 +668,30 @@ void NetworkOPsImp::submitTransaction (Job&, STTx::pointer iTrans)
     }
+    catch (...)
+    {
+        m_journal.warning << "Exception checking transaction" << txid;
```
`JLOG(m_journal.warning) <<`

ximinez replied:
Fixed.

vinniefalco commented on src/ripple/app/ledger/impl/LedgerConsensusImp.cpp:
```diff
@@ -44,6 +44,7 @@
 #include <ripple/overlay/predicates.h>
 #include <ripple/protocol/STValidation.h>
 #include <ripple/protocol/UintTypes.h>
+#include <ripple/protocol/st.h>
```
if you include `st.h` then you shouldn't need `STValidation.h` or `UintTypes.h`

ximinez replied:
Fixed.

vinniefalco commented on src/ripple/app/tx/impl/apply.cpp:
```diff
@@ -86,15 +93,79 @@ invoke_apply (ApplyContext& ctx)
+    if (!(flags & SF_SIGGOOD))
+    {
+        // Don't know signature state. Check it.
+        if (!tx.checkSign(allowMultiSign))
```
Should `checkSign` really be a member of `STTx`? That means it would have to call back into `src/ripple/app` to handle SusPays or include `apply.tx`, breaking levelization (`src/ripple/protocol` does not include anything in `src/ripple/app`).

ximinez replied:
`checkSign` could possibly be rewritten as a free function. `STTx` doesn't have many private members. That said, I think it should stay, partially because it fits logically, but also because my understanding of the SusPay hash/signature/proof/digest check could be _moved_ from SusPay.cpp to `STTx`. Obviously, I could be wrong, but a quick look at the three relevant fields: `sfMethod, sfDigest, sfProof` are unique to SusPays, and if so, then any misuse of them in another transaction could be treated as a signature failure.

vinniefalco replied:
That could be okay I think but you can't be including `apply.h` inside `src/ripple/protocol`

ximinez replied:
No, definitely not breaking levelization. Check out the latest commit `[EXPERIMENTAL/FOLD] Move SuspayFinish digest check to STTx.`. I'm not entirely happy with it, but that's the general idea.

ximinez commented:
I really wasn't happy with the `[EXPERIMENTAL/FOLD] Move SuspayFinish digest check to STTx.` commit that was here earlier, there are changes pending to that check logic in #1255, and it didn't really feel like it was in scope, so I removed the commit entirely to reduce the risk of lost changes.

The commit was cherry-picked into a new branch: https://github.com/ximinez/rippled/commits/SusPaySig.

There was a performance concern with the sig check being moved before preflight, but it's important to note that this behavior only really changes for transactions first seen during consensus.

vinniefalco commented:
:+1:

ximinez commented:
Rebased against `0.29.1-b9` in develop. Also pulled in @scottschurr's "Remove TxnSignApiFacade" from #1265 so I wouldn't have to deal with the conflicts later. Finally, added the last "Remove RIPPLE_ENABLE_MULTI_SIGN" commit to address the fact that multisign is now checked from `Rules` instead of a compiler flag.

JoelKatz commented:
:+1:

ximinez commented:
Squashed and rebased against `0.29.1-b10`.

vinniefalco commented on src/ripple/app/ledger/impl/LedgerConsensusImp.cpp:
```diff
@@ -1185,10 +1185,17 @@ void LedgerConsensusImp::accept (std::shared_ptr<SHAMap> set)
-            apply (accum, *item.second, tapNONE,
-                getApp().getHashRouter().sigVerify(),
+        {
+            if (checkValidity(getApp().getHashRouter(),
```
inject the hash router

ximinez replied:
done

vinniefalco commented on src/ripple/app/ledger/impl/LedgerConsensusImp.cpp:
```diff
@@ -1836,10 +1839,27 @@ applyTransaction (OpenView& view,
+        // in precheck.
+        assert(!(flags & tapENABLE_TESTING));
+        if ((*txn)[sfAccount] != zero &&
+            checkValidity(getApp().getHashRouter(),
```
inject

ximinez replied:
Done

vinniefalco commented on src/ripple/app/ledger/impl/LedgerMaster.cpp:
```diff
@@ -421,13 +421,13 @@ class LedgerMasterImp
-
+                    // To get into mHeldTransactions,
+                    // the signature must be valid.
+                    assert(checkValidity(getApp().getHashRouter(),
```
inject

ximinez replied:
Done

vinniefalco commented on src/ripple/app/main/Application.cpp:
```diff
@@ -1321,7 +1322,8 @@ bool ApplicationImp::loadOldLedger (
                     std::make_shared<Serializer const>(
                         std::move(s)), nullptr);
-                getApp().getHashRouter().setFlags (item.key(), SF_SIGGOOD);
+                forceValidity(getApp().getHashRouter(),
```
Are we calling `getApp` inside a member of `Application`?

JoelKatz replied:
We do that a lot. That's actually kind of funny.

ximinez replied:
I blame an overactive `Ctrl-V` keyboard shortcut.

ximinez replied:
Fixed

vinniefalco commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -654,52 +654,47 @@ void NetworkOPsImp::submitTransaction (STTx::pointer iTrans)
-    uint256 suppress = trans->getTransactionID ();
-    int flags;
+    auto const txid = trans->getTransactionID ();
+    auto const flags = getApp().getHashRouter().getFlags(txid);
```
Add `Application& app_` to `NetworkOPsImp`

vinniefalco commented on src/ripple/overlay/impl/PeerImp.cpp:
```diff
@@ -1054,10 +1056,34 @@ PeerImp::onMessage (std::shared_ptr <protocol::TMTransaction> const& m)
+        }
+        else
+        {
+            forceValidity(getApp().getHashRouter(),
```
Retrieve the Application object using `overlay_.app()`

ximinez commented:
Rebased onto PR #1289. Remaining calls to `getApp()` removed.

ximinez commented on src/ripple/app/ledger/impl/LedgerConsensusImp.cpp:
```diff
@@ -1826,10 +1833,6 @@ applyTransaction (Application& app, OpenView& view,
 
-    if ((app.getHashRouter ().getFlags (txn->getTransactionID ())
-            & SF_SIGGOOD) == SF_SIGGOOD)
-        flags = flags | tapNO_CHECK_SIGN;
```
I'm leaving this instance out because when the Transactor calls `checkValidity`, the same flags will get checked, so this instance is redundant.

ximinez commented on src/ripple/app/ledger/impl/LedgerMaster.cpp:
```diff
@@ -423,13 +423,13 @@ class LedgerMasterImp
                     ApplyFlags flags = tapNONE;
-                    if (app_.getHashRouter().addSuppressionFlags (
-                            it.first.getTXID (), SF_SIGGOOD))
-                        flags = flags | tapNO_CHECK_SIGN;
```
I'm leaving this instance out because when the Transactor calls checkValidity, the same flags will get checked, so this instance is redundant.

ximinez commented on src/ripple/app/ledger/impl/LedgerMaster.cpp:
```diff
@@ -443,14 +443,14 @@ class LedgerMasterImp
-                if (app_.getHashRouter ().addSuppressionFlags (
-                        it.first.getTXID (), SF_SIGGOOD))
-                    tepFlags = static_cast<ApplyFlags> (
-                        tepFlags | tapNO_CHECK_SIGN);
```
I'm leaving this instance out because when the Transactor calls checkValidity, the same flags will get checked, so this instance is redundant.

ximinez commented on src/ripple/app/ledger/impl/OpenLedger.cpp:
```diff
@@ -140,13 +143,22 @@ OpenLedger::apply_one (Application& app, OpenView& view,
-    if ((router.getFlags(
-        tx->getTransactionID()) & SF_SIGGOOD) ==
-            SF_SIGGOOD)
-        flags = flags | tapNO_CHECK_SIGN;
```
I'm leaving this instance out because when the Transactor calls checkValidity, the same flags will get checked, so this instance is redundant.

ximinez commented:
Rebased onto 0.29.1-b11.

I have selectively reverted changes related to `ApplyFlags::tapNO_CHECK_SIGN`. I left those changes separately as a fold commit, but it may be easier to see what's going on with the overall view (https://github.com/ripple/rippled/pull/1230/files). Initially this was done to prevent breaking changes introduced with the order book stuffing tests, but in retrospect, this may be the better solution anyway, as it leaves the callers to `ripple::apply` a lot cleaner, and removes the risk of a caller "forgetting" to check the signature, but avoids the crazy callbacks and injections that I was trying to avoid the first time around.

@JoelKatz, @miguelportilla, @vinniefalco I would appreciate if you'd give this another going over at your earliest convenience.

ximinez commented:
Rebased onto 0.29.1-b12. Sorry for the moving target, but there were conflicts.

ximinez commented:
Rebased yet again, this time against `0.29.1-rc1` and resolved conflicts.

vinniefalco commented on src/ripple/app/misc/HashRouter.h:
```diff
@@ -29,17 +29,20 @@
 #define SF_RELAYED      0x01    // Has already been relayed to other nodes
 // VFALCO NOTE How can both bad and good be set on a hash?
-#define SF_BAD          0x02    // Signature/format is bad
-#define SF_SIGGOOD      0x04    // Signature is good
```
How is it that we're removing `SF_SIGGOOD`??!

ximinez replied:
Not so much removing it as hiding it / making it private so that callers are forced to go through `checkValidity`:
* https://github.com/ripple/rippled/pull/1230/files#diff-a8d97ad254f542a49802c0397343ea2cR43
* https://github.com/ripple/rippled/pull/1230/files#diff-e785a4713dff55ac6cd5320a7b295956R41

vinniefalco commented on src/ripple/app/misc/HashRouter.h:
```diff
@@ -172,7 +162,7 @@ class HashRouter
     std::map< int, std::list<uint256> > mSuppressionTimes;
 
-    int mHoldTime;
+    int const mHoldTime;
```
If you make a data member const you should explicitly `delete` the copy assignment operator

ximinez replied:
Done.

vinniefalco commented on src/ripple/app/tests/HashRouter_test.cpp:
```diff
@@ -0,0 +1,184 @@
+//------------------------------------------------------------------------------
+/*
+This file is part of rippled: https://github.com/ripple/rippled
```
indentation is weird here

ximinez replied:
I blame VS auto-formatting. :)

vinniefalco commented on src/ripple/app/tests/HashRouter_test.cpp:
```diff
@@ -0,0 +1,184 @@
+        // key2 : 1
+        // key3 : null
+
+        UptimeTimer::getInstance().incrementElapsedTime();
```
ugh, this `UptimeTimer` thing...

vinniefalco commented on src/ripple/app/tests/Regression_test.cpp:
```diff
@@ -95,8 +94,7 @@ struct Regression_test : public beast::unit_test::suite
 
             auto const result = ripple::apply(env.app(),
```
Once a function receives `Application&` as a parameter, it makes no sense to also receive `Config const&` as a parameter. You could just call `app.config()`.

ximinez replied:
#1328 is a prerequisite for removing `config` from `apply`, so I'll get back to this one.

ximinez replied:
Done.

vinniefalco commented on src/ripple/app/tx/Transaction.h:
```diff
@@ -65,11 +64,11 @@ class Transaction
     static
     Transaction::pointer
-    sharedTransaction (Blob const&, Validate, Application& app);
+    sharedTransaction (Blob const&, Rules const& rules, Application& app);
```
I hate that we're building up code in this deprecated `Transaction` class

ximinez replied:
Just enough to maintain compatibility.

vinniefalco commented on src/ripple/app/tx/Transaction.h:
```diff
@@ -77,15 +76,20 @@ class Transaction
+    static
+    Transaction::pointer
+    transactionFromSQLValidated (
+        boost::optional<std::uint64_t> const& ledgerSeq,
```
Actually I think `ledgerSeq` should be `boost::optional<LedgerIndex> const& ledgerSeq`, and note that `LedgerIndex` is `std::uint32_t`.

ximinez replied:
The `Transactions.LedgerSeq` column is defined as a `BIGINT UNSIGNED`, and is being read as a `std::uint64_t`, and I ain't touching that, at least not in this PR.

vinniefalco commented on src/ripple/app/tx/apply.h:
```diff
@@ -31,6 +31,47 @@
+};
+
+/**
+Checks transaction signature and local checks. Returns
```
Wrong format and indentation

ximinez replied:
Fixed.

vinniefalco commented on src/ripple/overlay/impl/PeerImp.cpp:
```diff
@@ -1055,10 +1057,34 @@ PeerImp::onMessage (std::shared_ptr <protocol::TMTransaction> const& m)
+        if (checkSignature)
+        {
+            // Check the signature before handing off to the job queue.
+            auto valid = checkValidity(app_.getHashRouter(), *stx,
```
err... we can't be checking the signature on an `io_service` thread! I believe the signature needs to be checked from the `JobQueue`, so the load can be tracked and managed.

ximinez replied:
Fixed. Per discussion, this was originally a request to avoid the overhead of spinning up another job, but as we discussed, the cost of checking a signature is much higher.

vinniefalco commented on src/ripple/test/jtx/impl/Env.cpp:
```diff
@@ -248,14 +249,33 @@ Env::submit (JTx const& jt)
     {
+        // Use a local HashRouter object to avoid cross-caching
+        // the results of other tests.
+        // TODO: Get rid of this once each Env instantiates its own app()
```
This can be arranged

ximinez replied:
Done.

ximinez commented:
Rebased against Vinnie's #1328, which gives each test a separate `Application` object, and allows me to clean up some redundant `Config`s and remove some sig checking workarounds.

vinniefalco commented on src/ripple/overlay/impl/PeerImp.cpp:
```diff
@@ -1071,9 +1073,10 @@ PeerImp::onMessage (std::shared_ptr <protocol::TMTransaction> const& m)
             app_.getJobQueue ().addJob (
                 jtTRANSACTION, "recvTransaction->checkTransaction",
-                [weak, flags, stx] (Job&) {
+                [weak, flags, checkSignature, stx] (Job&) {
```
I believe this could be written
```
[weak = shared_from_this(), flags, checkSignature, stx](Job&)
```

ximinez replied:
That doesn't work because then `weak` gets `auto`-declared as a `shared_ptr` instead of a `weak_ptr`.

HowardHinnant replied:
We currently *allow* `RIPPLED_USE_CPP_14=1`.  With this change we would *require* `RIPPLED_USE_CPP_14=1`.  I'm ok with that.  I'm testing develop with that now just to double check it.  I'm just saying this to ensure that we don't make this happy and momentous move accidentally, but with eyes wide open.

ximinez replied:
You could write it as
```[weak = std::weak_ptr<PeerImp>(shared_from_this()), flags, checkSignature, stx] (Job&)```
but that looks ugly to me.

vinniefalco replied:
@HowardHinnant We dont need to set a macro just to use C++14 lambdas.. do we? Does that control the SConstruct?

vinniefalco replied:
@ximinez It might look a bit ugly but its definitely preferred because 1. It reduces the scope of `weak` to the lambda and 2. It avoids a needless move or copy

HowardHinnant replied:
Right.  That instructs SConstruct to use `-std=c++14` instead of `-std=c++11`.  And we need `-std=c++14` for this language feature (else we fail to compile).

HowardHinnant replied:
Develop compiles without warnings on OS X with `-std=c++14`.  Passes tests, and even syncs and stops on hotel wi-fi! :-)

vinniefalco replied:
@HowardHinnant `RIPPLED_USE_CPP_14` has no effect on Visual Studio builds (other than to change the contents of some beast headers)

HowardHinnant replied:
Not surprising.  VS doesn't really adhere to any particular C++ standard.  Instead it implements whatever features it can and throws them into its latest version.  But gcc and clang strictly implement C++11 and C++14 features, activated only under the `-std=c++11` / `-std=c++14` flags respectively.  It maintains these as two distinct languages forever.

ximinez replied:
I've pushed a commit making the change. If it works for @HowardHinnant and the CIs, I can squash it.

ximinez commented:
Well, CI checks passed with the updated lambda. Good to go?

vinniefalco commented:
It looks good to me, @HowardHinnant ?

HowardHinnant commented:
Visual inspection looks good to me.  If the compiler likes it, then :+1:.  Do we need to document somewhere that we now require `RIPPLED_USE_CPP_14=1`?

vinniefalco commented:
@HowardHinnant But it passed CI without setting `RIPPLED_USE_CPP_14=1`..

HowardHinnant commented:
Checking out to investigate...

HowardHinnant commented:
Without `RIPPLED_USE_CPP_14=1` clang issues two warnings:
````
clang++ ... -std=c++11 ...
src/ripple/overlay/impl/PeerSet.cpp
src/ripple/overlay/impl/PeerImp.cpp:1075:18: warning: initialized lambda
captures are a C++14 extension
      [-Wc++14-extensions]
                [weak = std::weak_ptr<PeerImp>(shared_from_this()),
                 ^
src/ripple/overlay/impl/PeerImp.cpp:1736:21: warning: add explicit braces to
avoid dangling else [-Wdangling-else]
                    JLOG(p_journal_.trace) <<
                    ^
src/ripple/basics/Log.h:225:29: note: expanded from macro 'JLOG'
#define JLOG(x) if (!x) { } else x
                            ^
````
`RIPPLED_USE_CPP_14=1` gets rid of the first warning.  Wrapping the indicated `JLOG` in `{}` gets rid of the second warning.

gcc also gives the first warning without `RIPPLED_USE_CPP_14=1`.

ximinez commented:
Hm. I forgot to check warnings in the CI. They're in there, too - both gcc (https://travis-ci.org/ripple/rippled/jobs/84120243) and clang (https://travis-ci.org/ripple/rippled/jobs/84120245).  So unless we make `RIPPLED_USE_CPP_14=1` the default, this isn't ready for prime-time.

There are also a couple of unused variable warnings in the Transactors, which I should clean up.

ximinez commented:
Pushed a commit to fix the gcc and clang warnings **unrelated** to C++14.

vinniefalco commented:
:+1:

nbougalis commented:
Merged as 66b55f91baf334d1ecdc1387df2b4145661f3058 and 9154cbf8e10613e67be8bb172ad0e64d51880c06.
